### PR TITLE
[6.4.r1] Import Samsung's exFAT filesystem kernel driver

### PR DESCRIPTION
**Pull request by artem**
Extracted from Samsung Galaxy S8+ SM-G955F open source code release for firmware version G955FXXU2CRED with additional patches.

### CODE REVIEW

**Comment by artem** [comment]
Samsung Galaxy S9 ships with k4.9 and has sdFAT version 2.0.8 that seems compatible with k4.4. Should I merge it or we are going to stick with 1.4.18?

**Comment by jerpelea** [comment]
use the new driver

File: fs/sdfat/sdfat.c [hunk 1]
```diff
@@ -4966,7 +4966,7 @@ static void sdfat_debug_kill_sb(struct super_block *sb)
 
 static struct file_system_type sdfat_fs_type = {
 	.owner       = THIS_MODULE,
-	.name        = "sdfat",
+	.name        = "exfat",
```
Comment (luk1337) [inline fs/sdfat/sdfat.c]: no.

File: fs/sdfat/core_fat.c [hunk 2]
```diff
@@ -0,0 +1,1463 @@
...
+
+	.get_au_stat = amap_get_au_stat,
+};
+
+s32 mount_fat16(struct super_block *sb, pbr_t *p_pbr)
+{
+	s32 num_root_sectors;
+	bpb16_t *p_bpb = &(p_pbr->bpb.f16);
+	FS_INFO_T *fsi = &(SDFAT_SB(sb)->fsi);
+
```
Comment (kholk) [inline fs/sdfat/core_fat.c]: #ifdef CONFIG_SDFAT_DISALLOW_FAT16
return -ENXIO;
#endif

File: fs/sdfat/core_fat.c [hunk 3]
```diff
@@ -0,0 +1,1463 @@
...
+	sdfat_log_msg(sb, KERN_INFO, "start_sect of partition  : %lld",
+		(s64)hidden);
+	return hidden;
+
+}
+
+s32 mount_fat32(struct super_block *sb, pbr_t *p_pbr)
+{
+	pbr32_t *p_bpb = (pbr32_t *)p_pbr;
+	FS_INFO_T *fsi = &(SDFAT_SB(sb)->fsi);
```
Comment (kholk) [inline fs/sdfat/core_fat.c]: #ifdef CONFIG_SDFAT_DISALLOW_FAT32
return -ENXIO;
#endif

**Comment by kholk** [comment]
Please make a commit with the requested changes.

Also, please add the suggested configuration options in the driver's Kconfig as DEFAULT Y (so, default disallowed). 

Thanks.

**Comment by kholk** [comment]
P.S.: Sorry about the bad indentation, writing from phone doesn't help!

**Review by kholk — CHANGES_REQUESTED** [review]

**Comment by danielgusvt** [comment]
@ab123321 is adding the old driver and after that adding a update commit actually necessary? Would it be possible to import only one commit for the driver?

**Comment by jerpelea** [comment]
[    1.043772] Registering sdcardfs 0.1
[   11.791278] sdcardfs version 2.0
[   11.791290] sdcardfs: dev_name -> /data/media
[   11.791294] sdcardfs: options -> fsuid=1023,fsgid=1023,multiuser,derive_gid,mask=6,userid=0,gid=1015
[   11.791298] sdcardfs: mnt -> ffffffc0d45cf820
[   11.791333] sdcardfs: mounted on top of /data/media type ext4
[   11.791998] sdcardfs : options - debug:1
[   11.792000] sdcardfs : options - gid:9997
[   11.792002] sdcardfs : options - mask:23
[   11.792535] sdcardfs : options - debug:1
[   11.792537] sdcardfs : options - gid:9997
[   11.792539] sdcardfs : options - mask:7

**Pull request merged**